File: 0111-minimum-depth-of-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        """
        1. traverse tree by level
        2. increment the ans each level
        3. exit as soon as a root leaf is found        
        """

        if root is None:
            return 0

        ans : int = 0
        q : List[TreeNode] = []
        q.append(root)

        while q:
            ans += 1
            size : int = len(q)

            for _ in range(size):
                
                node : TreeNode = q[0]
                q = q[1:]

                if node.left is None and node.right is None:
                    return ans
                
                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)


        return ans

File: 0111-minimum-depth-of-binary-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func minDepth(root *TreeNode) int {
	/*
    1. traverse tree by level
    2. increment the ans each level
    3. exit as soon as a root leaf is found
	*/

	ans := 0

	if root == nil {
		return ans
	}

	q := []*TreeNode{}
	q = append(q, root)

	for len(q) > 0 {
		ans += 1
		size := len(q)

		for i := 0; i < size; i++ {
            node := q[0]
            q = q[1:]

            if node.Left == nil && node.Right == nil {
				return ans
			}
			if node.Left != nil {
				q = append(q, node.Left)
			}
			if node.Right != nil {
				q = append(q, node.Right)
			}
		}

	}

	return ans
}
